inc/admin-layouts.php

```diff
@@ -0,0 +1,313 @@
+				wp_die();
+			}
+			
+			$layout = $layouts[ $_REQUEST['lid'] ];
+			
+			// Fix the format of this layout
+			if( !empty( $layout[ 'filename' ] ) ) {
+				$filename = $layout[ 'filename' ];
+				// Only accept filenames that end with .json
+				if( substr( $filename, -5, 5 ) === '.json' && file_exists( $filename ) ) {
+					$panels_data = json_decode( file_get_contents( $filename ), true );
+					$layout[ 'widgets' ] = ! empty( $panels_data[ 'widgets' ] ) ? $panels_data[ 'widgets' ] : array();
+					$layout[ 'grids' ] = ! empty( $panels_data[ 'grids' ] ) ? $panels_data[ 'grids' ] : array();
+					$layout[ 'grid_cells' ] = ! empty( $panels_data[ 'grid_cells' ] ) ? $panels_data[ 'grid_cells' ] : array();
+				}
+			}
+			
+			// A theme or plugin could use this to change the data in the layout
+			$panels_data = apply_filters( 'siteorigin_panels_prebuilt_layout', $layout, $lid );
+			
+			// Remove all the layout specific attributes
+			if ( isset( $panels_data['name'] ) ) unset( $panels_data['name'] );
+			if ( isset( $panels_data['screenshot'] ) ) unset( $panels_data['screenshot'] );
+			if ( isset( $panels_data['filename'] ) ) unset( $panels_data['filename'] );
+			
+		} elseif ( substr( $_REQUEST['type'], 0, 10 ) == 'directory-' ) {
+			$directory_id = str_replace( 'directory-', '', $_REQUEST['type'] );
+			$directories = SiteOrigin_Panels_Admin_Layouts::single()->get_directories();
+			$directory = ! empty( $directories[ $directory_id ] ) ? $directories[ $directory_id ] : false;
+			
+			if( ! empty( $directory ) ) {
+				$url = $directory[ 'url' ] . 'layout/' . urlencode( $_REQUEST[ 'lid' ] ) . '/?action=download';
+				if( ! empty( $directory[ 'args' ] ) && is_array( $directory[ 'args' ] ) ) {
+					$url = add_query_arg( $directory[ 'args' ], $url );
+				}
+				
+				$response = wp_remote_get( $url );
+				if ( $response['response']['code'] == 200 ) {
+					// For now, we'll just pretend to load this
+					$panels_data = json_decode( $response['body'], true );
+				} else {
+					// Display some sort of error message
+				}
+			}
+			
+		} elseif ( current_user_can( 'edit_post', $_REQUEST['lid'] ) ) {
+			$panels_data = get_post_meta( $_REQUEST['lid'], 'panels_data', true );
+		}
+		
+		$panels_data = apply_filters( 'siteorigin_panels_data', $panels_data );
+		
+		$panels_data['widgets'] = SiteOrigin_Panels_Admin::single()->process_raw_widgets( $panels_data['widgets'], array(), true, true );
+		
+		echo json_encode( $panels_data );
+		wp_die();
+	}
+	
+	/**
+	 * Ajax handler to import a layout
+	 */
+	function action_import_layout() {
+		if ( empty( $_REQUEST['_panelsnonce'] ) || ! wp_verify_nonce( $_REQUEST['_panelsnonce'], 'panels_action' ) ) {
+			wp_die();
+		}
+		
+		if ( ! empty( $_FILES['panels_import_data']['tmp_name'] ) ) {
+			header( 'content-type:application/json' );
+			$json = file_get_contents( $_FILES['panels_import_data']['tmp_name'] );
+			@unlink( $_FILES['panels_import_data']['tmp_name'] );
```

Comment:
Yeah, that's a good idea. Will save that for a future enhancement though - https://github.com/siteorigin/siteorigin-panels/issues/388